File: cxx/Application.cxx
#include "include/Application.h"

#include "quickfix-rs/src/lib.rs.h"

Application::Application(): inner(new_application_impl()) {
  const auto session_id = FIX::SessionID();
  this->onCreate(session_id);
}

Application::~Application() {}

auto Application::onCreate(const FIX::SessionID& session_id) -> void {
  this->inner->on_create(session_id);
}

auto Application::onLogon(const FIX::SessionID& session_id) -> void {}

auto Application::onLogout(const FIX::SessionID& session_id) -> void {}

auto Application::toAdmin(FIX::Message& message, const FIX::SessionID& session_id) -> void {}

auto Application::fromAdmin(const FIX::Message& message, const FIX::SessionID& session_id) -> void {}

auto Application::toApp(FIX::Message& message, const FIX::SessionID& session_id) -> void {}

auto Application::fromApp(const FIX::Message& message, const FIX::SessionID& session_id) -> void {}

auto new_application() -> std::unique_ptr<Application> {
  return std::unique_ptr<Application>(new Application());
}

File: src/application.rs
use crate::ffi::{Message, SessionID};

pub struct ApplicationImpl {}

pub fn new_application_impl() -> Box<ApplicationImpl> {
    Box::new(ApplicationImpl {})
}

impl ApplicationImpl {
    pub fn on_create(&self, session_id: &SessionID) {
        println!("on_create {}", session_id.toStringFrozen());
    }

    pub fn on_logon(&self, session_id: &SessionID) {
        println!("on_logon {}", session_id.toStringFrozen());
    }

    pub fn on_logout(&self, session_id: &SessionID) {
        println!("on_logout {}", session_id.toStringFrozen());
    }

    pub fn to_admin(&self, message: &mut Message, session_id: &SessionID) {
        todo!()
    }

    pub fn from_admin(&self, message: &Message, session_id: &SessionID) {
        todo!()
    }

    pub fn to_app(&self, message: &mut Message, session_id: &SessionID) {
        todo!()
    }

    pub fn from_app(&self, message: &Message, session_id: &SessionID) {
        todo!()
    }
}
